### PR TITLE
Implement Shipping Method Index/Create/Update pages

### PR DESCRIPTION
**Pull request by jakubtobiasz**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | `bootstrap-admin-panel`

### CODE REVIEW

**Review by mpysiak — APPROVED** [review]

File: src/Sylius/Bundle/AdminBundle/Resources/views/Shared/form_theme.html.twig [hunk 1]
```diff
@@ -25,6 +25,64 @@
     </div>
 {% endblock %}
 
+{% block sylius_shipping_method_rule_row %}
```
Comment (Wojdylak) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Shared/form_theme.html.twig]: IMO it is better to keep it in `ShippingMethod/form_theme.html.twig` file

File: src/Sylius/Bundle/AdminBundle/Resources/views/Shared/form_theme.html.twig [hunk 2]
```diff
@@ -25,6 +25,64 @@
...
+{% block sylius_promotion_rule_row %}
+    <div {% with {attr: row_attr|merge({class: row_attr.class|default('mb-3 card')|trim})} %}{{ block('attributes') }}{% endwith %}>
+        <div class="card-body" {{ sylius_test_html_attribute('promotion-rule') }}>
+            {{- form_label(form) -}}
+            {{- form_widget(form) -}}
+            {{- form_errors(form) -}}
+            {{- form_row(button_delete, sylius_test_form_attribute('delete-rule')) -}}
+        </div>
+    </div>
+{% endblock %}
```
Comment (Wojdylak) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Shared/form_theme.html.twig]: This has been moved to `promotion.form_theme.html.twig`

File: src/Sylius/Bundle/AdminBundle/Resources/views/Shared/form_theme.html.twig [hunk 3]
```diff
@@ -25,6 +25,64 @@
     </div>
 {% endblock %}
 
+{% block sylius_shipping_method_rule_row %}
+    <div{% with {attr: row_attr|merge({class: row_attr.class|default('mb-3 card')|trim})} %}{{ block('attributes') }}{% endwith %} {{ sylius_test_html_attribute('rule') }}>
+        <div class="card-body">
+            {{- form_label(form) -}}
+            {{- form_widget(form) -}}
+            {{- form_errors(form) -}}
```
Comment (Wojdylak) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Shared/form_theme.html.twig]: There isn't button to delete the rule :stuck_out_tongue:

File: src/Sylius/Bundle/ShippingBundle/Form/Type/ShippingMethodType.php [hunk 4]
```diff
@@ -70,15 +70,11 @@ public function buildForm(FormBuilderInterface $builder, array $options): void
             ])
             ->add('calculator', CalculatorChoiceType::class, [
                 'label' => 'sylius.form.shipping_method.calculator',
+                'placeholder' => 'sylius.form.shipping_method.select_calculator',
             ])
             ->add('enabled', CheckboxType::class, [
                 'label' => 'sylius.form.locale.enabled',
             ])
-            ->add('rules', ShippingMethodRuleCollectionType::class, [
```
Comment (Wojdylak) [inline src/Sylius/Bundle/ShippingBundle/Form/Type/ShippingMethodType.php]: Why did you delete it?

File: src/Sylius/Bundle/AdminBundle/TwigComponent/ShippingMethod/FormComponent.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    public function dehydrateResource(): ?int
+    {
+        return $this->resource?->getId();
+    }
+
+    /**
+     * @return array<string, mixed>
+     */
+    #[ExposeInTemplate(name: 'hookable_data')]
+    public function getHookableData(): array
```
Comment (Wojdylak) [inline src/Sylius/Bundle/AdminBundle/TwigComponent/ShippingMethod/FormComponent.php]: Why do you need this?

File: src/Sylius/Bundle/AdminBundle/Resources/views/Shared/form_theme.html.twig [hunk 6]
```diff
@@ -25,6 +25,64 @@
...
+
+{% block sylius_shipping_method_rule_total_weight_greater_than_or_equal_configuration_row %}
+    <div
+        {% with {attr: row_attr|merge({class: row_attr.class|default('mb-3 card')|trim})} %}{{ block('attributes') }}{% endwith %}
+        {{ sylius_test_html_attribute('shipping-method-weight-rule') }}
+    >
+        <div class="card-body">
+            {{- form_label(form) -}}
+            {{- form_widget(form) -}}
+            {{- form_errors(form) -}}
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Shared/form_theme.html.twig]: Adding `form_help` by default would be nice, but we can just adjust it globally later

File: src/Sylius/Bundle/AdminBundle/Resources/views/ShippingMethod/form.html.twig [hunk 7]
```diff
@@ -0,0 +1,36 @@
+{% set main_hook = hook_name(hookable_data.parent_main_hook, 'form') %}
+{% set fallback_hook = hook_name(hookable_data.parent_fallback_hook, 'form') %}
+{% set form = hookable_data.form %}
+
+{% form_theme form '@SyliusAdmin/Shared/form_theme.html.twig' %}
+
+<div class="container-xl" {{ attributes }}>
+    {{ form_start(form, {'attr': {'class': 'ui loadable form', 'novalidate': 'novalidate', 'id': form.vars.id}}) }}
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/AdminBundle/Resources/views/ShippingMethod/form.html.twig]:
```suggestion
    {{ form_start(form, {'attr': {'class': 'ui loadable form', 'novalidate': 'novalidate'}}) }}
```
Pretty sure `id` should be there without outright passing it.
Also the `class` attribute looks like semantic-ui, is it correct?

File: src/Sylius/Bundle/ShippingBundle/Form/Type/ShippingMethodRuleCollectionType.php [hunk 8]
```diff
@@ -1,32 +0,0 @@
...
- */
-
-declare(strict_types=1);
-
-namespace Sylius\Bundle\ShippingBundle\Form\Type;
-
-use Sylius\Bundle\ShippingBundle\Form\Type\Core\AbstractConfigurationCollectionType;
-use Symfony\Component\OptionsResolver\OptionsResolver;
-
-final class ShippingMethodRuleCollectionType extends AbstractConfigurationCollectionType
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ShippingBundle/Form/Type/ShippingMethodRuleCollectionType.php]: Should be reverted

File: src/Sylius/Bundle/ShippingBundle/Resources/config/services/form.xml [hunk 9]
```diff
@@ -67,11 +67,6 @@
             <tag name="form.type" />
         </service>
 
-        <service id="sylius.form.type.shipping_method_rule.collection" class="Sylius\Bundle\ShippingBundle\Form\Type\ShippingMethodRuleCollectionType">
```
Comment (NoResponseMate) [inline src/Sylius/Bundle/ShippingBundle/Resources/config/services/form.xml]: Should be reverted

File: src/Sylius/Bundle/AdminBundle/Resources/views/ShippingMethod/form.html.twig (hunk 7 above)
Comment (Wojdylak) [reply src/Sylius/Bundle/AdminBundle/Resources/views/ShippingMethod/form.html.twig]: I have removed the `class` attribute, but have left the `id` attribute, as it is not passed by default.

File: src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php [hunk 10]
```diff
@@ -403,14 +403,15 @@ public function iShouldSeeShippingMethodsInTheList(int $count): void
...
 
     /**
-     * @Then the shipping method :name should be in the registry
-     * @Then the shipping method :name should appear in the registry
+     * @Then the shipping method :shippingMethod should be in the registry
+     * @Then the shipping method :shippingMethod should appear in the registry
+     * @Then the :shippingMethod shipping method should be successfully created
      */
-    public function theShippingMethodShouldAppearInTheRegistry(string $name): void
+    public function theShippingMethodShouldAppearInTheRegistry(ShippingMethodInterface $shippingMethod): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingShippingMethodsContext.php]: if nothing more than the name is needed in this method, then this change is not necessary, because this way we additionally query the database

File: src/Sylius/Behat/Context/Ui/Admin/ManagingShippingMethodsContext.php [hunk 11]
```diff
@@ -14,34 +14,36 @@
...
 {
-    use ValidationTrait;
-
     public function __construct(
         private IndexPageInterface $indexPage,
         private CreatePageInterface $createPage,
         private UpdatePageInterface $updatePage,
-        private CurrentPageResolverInterface $currentPageResolver,
-        private NotificationCheckerInterface $notificationChecker,
+        private FormElementInterface $shippingMethodForm,
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingShippingMethodsContext.php]:
```suggestion
        private FormElementInterface $shippingMethodFormElement,
```
? 🤔

File: src/Sylius/Behat/Page/Admin/Crud/IndexPage.php [hunk 12]
```diff
@@ -129,6 +129,22 @@ public function filter(): void
...
+
+            return null !== $accordionCollapse && !$accordionCollapse->hasClass('collapsing');
+        });
+    }
+
+    protected function areFiltersVisible(): bool
+    {
+        return !$this->getElement('filters_toggle')->hasClass('collapsed');
+    }
+
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Admin/Crud/IndexPage.php]: Protected methods should be after the public ones

File: src/Sylius/Bundle/AdminBundle/Resources/config/services/form.xml [hunk 13]
```diff
@@ -43,6 +43,15 @@
             <tag name="form.type_extension" extended-type="Sylius\Bundle\AddressingBundle\Form\Type\CountryType" priority="100" />
         </service>
 
+        <service id="sylius_admin.form.extension.shipping_method_type" class="Sylius\Bundle\AdminBundle\Form\Extension\ShippingMethodTypeExtension">
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/config/services/form.xml]:
```suggestion
        <service id="sylius_admin.form.extension.shipping_method" class="Sylius\Bundle\AdminBundle\Form\Extension\ShippingMethodTypeExtension">
```

**Review by GSadee — APPROVED** [review]

File: src/Sylius/Behat/Context/Ui/Admin/ManagingShippingMethodsContext.php [hunk 14]
```diff
@@ -14,34 +14,36 @@
...
 {
-    use ValidationTrait;
-
     public function __construct(
         private IndexPageInterface $indexPage,
         private CreatePageInterface $createPage,
         private UpdatePageInterface $updatePage,
-        private CurrentPageResolverInterface $currentPageResolver,
-        private NotificationCheckerInterface $notificationChecker,
+        private FormElementInterface $shippingMethodFormElement,
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingShippingMethodsContext.php:38]:
```suggestion
        private FormElementInterface $formElement,
```
🤔

File: src/Sylius/Behat/Context/Ui/Admin/ManagingShippingMethodsContext.php [hunk 15]
```diff
@@ -239,11 +248,25 @@ public function iSaveMyChanges()
...
+        Assert::same(
+            $this->shippingMethodFormElement->getValidationMessage($field),
+            sprintf('Please enter shipping method %s.', $field),
+        );
+    }
+
+    /**
+     * @Then I should be notified that name is required
+     */
+    public function iShouldBeNotifiedThatNameIsRequired($localeCode = 'en_US'): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingShippingMethodsContext.php:264]:
```suggestion
    public function iShouldBeNotifiedThatNameIsRequired(string $localeCode = 'en_US'): void
```

File: src/Sylius/Behat/Context/Ui/Admin/ManagingShippingMethodsContext.php [hunk 16]
```diff
@@ -239,11 +248,25 @@ public function iSaveMyChanges()
...
     {
-        $this->assertFieldValidationMessage($element, sprintf('Please enter shipping method %s.', $element));
+        Assert::same(
+            $this->shippingMethodFormElement->getValidationMessage($field),
+            sprintf('Please enter shipping method %s.', $field),
+        );
+    }
+
+    /**
+     * @Then I should be notified that name is required
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingShippingMethodsContext.php:262]: `Then` methods should be after `When` ones

File: src/Sylius/Behat/Page/Admin/ShippingMethod/CreatePage.php [hunk 17]
```diff
@@ -13,179 +13,8 @@
...
-use Behat\Mink\Element\NodeElement;
-use Behat\Mink\Exception\ElementNotFoundException;
-use Sylius\Behat\Behaviour\SpecifiesItsField;
 use Sylius\Behat\Page\Admin\Crud\CreatePage as BaseCreatePage;
-use Sylius\Behat\Service\DriverHelper;
-use Sylius\Behat\Service\TabsHelper;
-use Sylius\Component\Core\Formatter\StringInflector;
-use Webmozart\Assert\Assert;
 
 class CreatePage extends BaseCreatePage implements CreatePageInterface
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Admin/ShippingMethod/CreatePage.php:18]: Both `CreatePage` and `UpdatePage` for shipping method could be probably removed as they are empty now

File: src/Sylius/Bundle/AdminBundle/Resources/config/services/form.xml [hunk 18]
```diff
@@ -43,6 +43,15 @@
             <tag name="form.type_extension" extended-type="Sylius\Bundle\AddressingBundle\Form\Type\CountryType" priority="100" />
         </service>
 
+        <service id="sylius_admin.form.extension.shipping_method" class="Sylius\Bundle\AdminBundle\Form\Extension\ShippingMethodTypeExtension">
+            <argument>%sylius.shipping_method_rules%</argument>
+            <tag name="form.type_extension" />
+        </service>
+
+        <service id="sylius.form.extension.shipping_method_rule" class="Sylius\Bundle\AdminBundle\Form\Extension\ShippingMethodRuleTypeExtension">
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/config/services/form.xml:51]:
```suggestion
        <service id="sylius_admin.form.extension.shipping_method_rule" class="Sylius\Bundle\AdminBundle\Form\Extension\ShippingMethodRuleTypeExtension">
```

**Review by GSadee — APPROVED** [review]
